File: resources/exercises/recursion/cons/ismember/IsMemberEntity.java
package recursion.cons.ismember;

import plm.universe.bat.BatTest;
import plm.universe.cons.ConsEntity;
import plm.universe.cons.RecList;

public class IsMemberEntity extends ConsEntity {

	@Override
	public void run(BatTest t) {
		RecList parameter = RecList.fromArray( (int[]) t.getParameter(0) );
		t.setResult( isMember( parameter, (int)t.getParameter(1) ) );
	}

	/* BEGIN TEMPLATE */
	boolean isMember(RecList seq, int val) {
		/* BEGIN SOLUTION */
		if (seq == null)
			return false;
		if (seq.head == val)
			return true;
		return isMember(seq.tail,val);
		/* END SOLUTION */
	}
	/* END TEMPLATE */
}

File: src/lessons/recursion/cons/IsMemberEntity.scala
package lessons.recursion.cons

import plm.universe.bat.BatEntity
import plm.universe.bat.BatTest
import lessons.recursion.cons.universe.ConsEntity

class IsMemberEntity extends ConsEntity {

    override def run() {
      import plm.core.ValueSerializer._

      val count = getTestCount()
      for (i <- 0 to count -1) {
        val param = deserialize(getTest(i)).asInstanceOf[Array[Object]]
      	setTestResult(i, serialize(isMember( param(0).asInstanceOf[Array[Int]].toList, param(1).asInstanceOf[Int] ) ))
	  }
	}

	/* BEGIN TEMPLATE */
	def isMember(l:List[Int], v:Int): Boolean = {
	/* BEGIN SOLUTION */
  l match {
    case a::b if a==v => true
    case a::b         => isMember(b,v)
    case _    => false
  }
	/* END SOLUTION */
	}
	/* END TEMPLATE */
}
